### PR TITLE
backport of #13819: Speed seeding for HLT

### PR DESCRIPTION
**Pull request by fwyzard**

### CODE REVIEW

**Comment by fwyzard** [comment]
tracked at #15151

**Comment by fwyzard** [comment]
please test

**Comment by fwyzard** [comment]
Does not introduce any changes to HLT results over 1000 events:
```
Found 1000 matching events, out of which 0 have different HLT results
```

File: Geometry/CommonDetUnit/interface/GeomDet.h [hunk 1]
```diff
@@ -119,9 +125,10 @@ class GeomDet {
 
   ReferenceCountingPointer<Plane>  thePlane;
   DetId m_detId;
-  int m_index;
+  int m_index=-1;
+  int m_gdetIndex=-1;
 protected:
-  AlignmentPositionError*               theAlignmentPositionError;
+  AlignmentPositionError* theAlignmentPositionError=nullptr;
```
Comment (ianna) [inline Geometry/CommonDetUnit/interface/GeomDet.h]: @fwyzard - would using std::unique_ptr work here?

**Comment by fwyzard** [comment]
the failure seems to be due to a missing input sample:
```
----- Begin Fatal Exception 26-Jul-2016 15:04:35 CEST-----------------------
An exception of category 'FallbackFileOpenError' occurred while
   [0] Constructing the EventProcessor
   [1] Constructing input source of type PoolSource
   [2] Calling RootFileSequenceBase::initTheFile()
   [3] Calling StorageFactory::open()
   [4] Calling XrdFile::open()
Exception Message:
Failed to open the file 'root://xrootd-cms.infn.it//store/relval/CMSSW_7_6_0_pre4/RelValZEE_13/MINIAODSIM/PU25ns_76X_mcRun2_asymptotic_v1-v1/00000/BE21962F-DD4F-E511-B681-002354EF3BDF.root'
   Additional Info:
      [a] XrdCl::File::Open(name='root://eoscms.cern.ch//eos/cms/store/relval/CMSSW_7_6_0_pre4/RelValZEE_13/MINIAODSIM/PU25ns_76X_mcRun2_asymptotic_v1-v1/00000/BE21962F-DD4F-E511-B681-00235
' (errno=3011, code=400). No additional data servers were found.
      [b] Last URL tried: root://eoscms.cern.ch:1094//eos/cms/store/relval/CMSSW_7_6_0_pre4/RelValZEE_13/MINIAODSIM/PU25ns_76X_mcRun2_asymptotic_v1-v1/00000/BE21962F-DD4F-E511-B681-002354EF
      [c] Problematic data server: eoscms.cern.ch:1094
      [d] Disabled source: eoscms.cern.ch:1094
      [e] Input file root://eoscms.cern.ch//eos/cms/store/relval/CMSSW_7_6_0_pre4/RelValZEE_13/MINIAODSIM/PU25ns_76X_mcRun2_asymptotic_v1-v1/00000/BE21962F-DD4F-E511-B681-002354EF3BDF.root
Fallback Input file root://xrootd-cms.infn.it//store/relval/CMSSW_7_6_0_pre4/RelValZEE_13/MINIAODSIM/PU25ns_76X_mcRun2_asymptotic_v1-v1/00000/BE21962F-DD4F-E511-B681-002354EF3BDF.root also
Original exception info is above; fallback exception info is below.
      [f] XrdCl::File::Open(name='root://xrootd-cms.infn.it//store/relval/CMSSW_7_6_0_pre4/RelValZEE_13/MINIAODSIM/PU25ns_76X_mcRun2_asymptotic_v1-v1/00000/BE21962F-DD4F-E511-B681-002354EF3
' (errno=3011, code=400). No additional data servers were found.
      [g] Last URL tried: root://cms-xrd-transit.cern.ch:1094//store/relval/CMSSW_7_6_0_pre4/RelValZEE_13/MINIAODSIM/PU25ns_76X_mcRun2_asymptotic_v1-v1/00000/BE21962F-DD4F-E511-B681-002354E
      [h] Problematic data server: cms-xrd-transit.cern.ch:1094
      [i] Disabled source: cms-xrd-transit.cern.ch:1094
----- End Fatal Exception -------------------------------------------------
```

**Comment by fwyzard** [comment]
please test

**Comment by fwyzard** [comment]
rebased on top of CMSSW_8_0_16

**Comment by fwyzard** [comment]
please test

**Comment by cvuosalo** [comment]
@fwyzard: Reco would like the "-Ofast" compiler option removed from this PR. My tests show no clear benefit and only an increase of differences with the option.
I tested timing using workflow 25202.0_TTbar_13 with 150 events against baseline CMSSW_8_0_X_2016-07-24-2300. The results for key modules with and without "-Ofast" follow.
```
Top entry of each pair has no Ofast, lower entry is with Ofast

delta/mean delta/orJob     original                   new       module name
---------- ------------     --------                  ----       ------------
-1.122691      -0.34%        39.86 ms/ev ->        11.20 ms/ev ecalDrivenElectronSeeds
-1.118605      -0.34%        39.86 ms/ev ->        11.27 ms/ev ecalDrivenElectronSeeds

-0.384773      -0.02%        25.51 ms/ev ->        17.28 ms/ev hltIter1PFlowPixelSeeds
-0.366628      -0.02%        25.51 ms/ev ->        17.60 ms/ev hltIter1PFlowPixelSeeds

-0.414749      -0.01%        12.28 ms/ev ->         8.06 ms/ev hltIter2PFlowPixelSeeds
-0.409385      -0.01%        12.28 ms/ev ->         8.11 ms/ev hltIter2PFlowPixelSeeds

-0.402262      -0.00%         4.43 ms/ev ->         2.95 ms/ev hltIter1ElectronsPixelSeeds
-0.399818      -0.00%         4.43 ms/ev ->         2.95 ms/ev hltIter1ElectronsPixelSeeds

-0.368544      -0.00%         5.66 ms/ev ->         3.90 ms/ev hltIter1PFlowPixelSeedsForTau
-0.343869      -0.00%         5.66 ms/ev ->         4.00 ms/ev hltIter1PFlowPixelSeedsForTau

-0.183275      -0.00%         8.59 ms/ev ->         7.15 ms/ev hltIter2ElectronsPixelSeeds
-0.225945      -0.00%         8.59 ms/ev ->         6.84 ms/ev hltIter2ElectronsPixelSeeds

-0.094925      -0.01%        28.77 ms/ev ->        26.16 ms/ev hltEgammaElectronPixelSeeds
-0.122846      -0.01%        28.77 ms/ev ->        25.44 ms/ev hltEgammaElectronPixelSeeds
```
The timing with Ofast shows no consistent improvement.

Second, in measuring DQM differences form baseline, Ofast produces 165 DQM plots with differences, versus 91 plots without Ofast. These differences in the plots are tiny, but Ofast is still creating unnecessary jitter or minor discrepancies.
This PR produces significant timing improvements without creating significant differences, but Ofast does not provide any of this improvement, so it should be eliminated.

**Comment by fwyzard** [comment]
@cvuosalo I will not be able to repeat the HLT performance tests before next week, I'll trust you that -Ofast is not responsible for any significant gain.

Is it OK if I just revert it in the last commit ?

**Comment by fwyzard** [comment]
@cmsbuild, please test

**Comment by cvuosalo** [comment]
@fwyzard: Why did you change to "O2" optimization? I didn't test with that option, and it wasn't there previously.

**Comment by cvuosalo** [comment]
@cmsbuild please test
Maybe Jenkins tests are working now?

**Comment by fwyzard** [comment]
@cvuosalo you asked to revert -Ofast, which was introduced by 610aad633dd878e3f43dc1aee13f9a262e3d63f2 , and before -Ofast it was -O2.

**Comment by fwyzard** [comment]
I see now that originally -O2 was not present either.
Let me know if you would prefer that I remove it altogether.
Or, just make a PR to revert it once this is merged.

**Comment by cvuosalo** [comment]
@fwyzard: For simplicity, let's revert to the original state, which has no optimization option in the BuildFile.xml, so that file will not even be part of this PR.

**Comment by fwyzard** [comment]
...this is starting to become rather annoying...

**Comment by fwyzard** [comment]
history has been rewritten, purging all changes to RecoEgamma/EgammaElectronAlgos/BuildFile.xml

**Comment by fwyzard** [comment]
please test

**Comment by cvuosalo** [comment]
@cmsbuild please test
Has Jenkins been fixed yet?

**Comment by smuzaffar** [comment]
please test

**Comment by cvuosalo** [comment]
@cmsbuild please test
@smuzaffar says 80X Jenkins tests should complete now

**Comment by cvuosalo** [comment]
+1

For #15280 bb3c568ffb3206db2d76d9f8de6fefc46dabe3d1

Speeding up seeding for HLT.

The code changes are satisfactory, and Jenkins tests against baseline CMSSW_8_0_X_2016-08-08-2300 show a few, tiny, insignificant differences. Extended tests discussed above (https://github.com/cms-sw/cmssw/pull/15280#issuecomment-237699658) show numerous, tiny differences, but they are not significant nor more than jitter and minor fluctuations. Timing tests (https://github.com/cms-sw/cmssw/pull/15280#issuecomment-235231359) show this PR provides about a 1.2% improvement.

**Pull request merged**